add normalization to conic drawing

strobelm opened the pull request:
I investigated bug #102. The conic drawing routines runs in to trouble in some examples due to a lack of normalization. After normalizing the matrix in the routine the degenerate drawings vanish.

This should resolve #102, perhaps @gagern could verify.

gagern commented:
Yes, this does resolve #102.

I'm a bit unhappy about the location where you fixed this: in my opinion, the renderer should treat the geometry as read-only. Even though a change in representative should be meaningless in most situations, the fact that you added this to fix a bug indicates that reality is something different. And I could imagine that in a few years we might be spending days debugging code which behaves different depending on whether you draw things in between.

Points get normalized after essentially any geometric operation. Perhaps we should do the same for conics. Or if we normalize things in the renderer, then we should not write the normalized representation back to the element. I fear we might want both of these: the former will ensure that operations other than rendering won't suffer from overflow or underflow due to lack of normalization. And the latter will ensure that the fix also applies to conics drawn by the user using a script command instead of a gslp element. What do you think?

gagern commented:
I fear this issue was my fault. As I just noticed in https://github.com/CindyJS/CindyJS/pull/100#issuecomment-161142656, I incorrectly removed some normalization from `SelectConic` in the code I was using during testing. Changing the code from select-by-position to select-by-index, things look smooth against the current master.

So the original point of this ticket here seems to be moot. Nevertheless, there is this question of whether we should normalize input from scripts. Should I file a new report for this, or do you want to re-purpose this one here to that effect?

strobelm commented:
I'm fine reusing this one and agree that altering the element from the drawing probably isn't the best idea, i was a bit hasty. 

Normalizing the conic handed over by scripts is generally a good idea. The best thing would be that the drawing ops always get a normalized version. Nevertheless i'm currently in favor of a fallback, non element altering, in the drawing routine, just to be sure.

strobelm commented:
I updated the branch and added normalization to evaluator.drawconic, removed the modification of the element and changed some lines for performance in the drawing routine. 

the variable perc handles how many points are generated at the apex of the parabola, increasing this from 5 to 10% has not a real performance impact, but the conics look a bit better.

gagern commented on src/js/libcs/OpDrawing.js:
```diff
@@ -389,8 +391,8 @@ eval_helper.drawconic = function(aConic, modifs) {
-    var mat = aConic.matrix;
-    var origmat = aConic.matrix;
+    var mat = List.normalizeMax(aConic.matrix);
+    var origmat = JSON.parse(JSON.stringify(mat));
```
Why would you want to do that JSON detour? All our CindyScript-typed data structures are treated as immutable in any case, so there should be no reason to make a copy. The [conic transformation](https://github.com/strobelm/CindyJS/blob/927e87cd702d2f48693a1f5acded7ff70d413774/src/js/libcs/OpDrawing.js#L427-L429) creates a new matrix and causes `mat` to refer to this, without affecting the original matrix.

strobelm replied:
You are right, the unnecessary detour was due to safety considerations. I'll change that.

gagern commented on src/js/libcs/OpDrawing.js:
```diff
@@ -377,7 +377,9 @@ evaluator.drawconic$1 = function(args, modifs) {
 
+
     }
+    Conic.matrix = List.normalizeMax(Conic.matrix);
```
For a moment I was afraid that this, too, would modify an existing element. But that's just a local object. Which made me think: why do we have that object? Why not simply pass the matrix directly to `eval_helper.drawconic`, without enclosing it into such an object? Not really related to the issue at hand here, just got me wondering.

strobelm replied:
Currently eval_helper.drawconic relies on the fact that aConic is handed over as an object with property matrix, we could of course change that which would entail some reworking of calls of drawconic.

strobelm commented:
Here is the update. Let me know if there are further wishes (especially the re-working of the eval_helper.drawconic argument handling). If not i would squash the commits for merging this.

gagern commented:
Sorry, I forgot about this request. Yes, I'd say change the argument handling to just pass the matrix, then squash this.

strobelm commented:
Here is the squashed update with direct matrix handling.
